docs/kb/rules/ellipsis-metavariables.md

```diff
@@ -0,0 +1,40 @@
+---
+description: Ellipsis metavariables can help with matching multiple word tokens.
+tags:
+  - Rules
+  - Semgrep Code
+---
+
+import MoreHelp from "/src/components/MoreHelp"
+
+# Matching multiple tokens with ellipsis metavariables
+
```

Comment:
```suggestion
Match multiple tokens, such as arguments of a function or elements of an array, through the use of two Semgrep rule constructs:

* Ellipsis metavariables
* Aliengrep
```

Just adding an intro and roughly defining a "token". Your subsequent definition of a token is not replaced.

Reply:
This isn't something I want to have upfront, in particular because this KB is about a less typical usage of ellipsis metavariables that isn't either of those things. But you have very quickly identified something that definitely isn't clear, and most of your comments basically center on this: what is a token?

Most of our docs only talk about tokens in the context of authentication (which is a special use/type of token), so we don't have a lot more on this, except in a few places. The [caveats section](https://semgrep.dev/docs/writing-rules/generic-pattern-matching/#caveats-and-limitations-of-generic-mode) that I link to is somewhat helpful:

> metavariable support is limited to capturing a single “word”, which is a token of the form [A-Za-z0-9_]+

Tokenization in language processing in general is a phase of analysis where you separate the language into its more basic elements. The details of the process, and what constitutes a token, depend on the language being analyzed and the purpose of the analysis. In the case of the generic language processor for Semgrep, it tokenizes `hello, world` as 3 tokens (as in the example), but only two of those tokens are "words"; the comma is a separator/punctuation. `ABC-DEF` is also broken down into 3 (two "words", ABC and DEF, and one separator, `-`), but not ABC_DEF - that's one token, because underscore is considered part of a word in that processor.

@emjin might have a good way to break this down more accessibly so we can present a brief definition that makes sense...

Reply:
> This isn't something I want to have upfront, in particular because this KB is about a less typical usage of ellipsis metavariables that isn't either of those things.

Yeah, I guess the array / function example doesn't match the rest of the article. What would make a better example or use case in this case for generic "words"?

Reply:
I ended up reorganizing the content at the top and adding more clear definitions/examples of terms to try to resolve this. Hopefully it make sense. :)